test/conicinterface.jl

```diff
@@ -196,16 +196,157 @@ MathProgBase.loadconicproblem!(m,
 MathProgBase.optimize!(m)
 @test MathProgBase.status(m) == :Optimal
-@test_approx_eq_eps MathProgBase.getobjval(m) -sqrt(2.0) 1e-6
-@test_approx_eq_eps MathProgBase.getsolution(m)[1] 1.0 1e-6
-@test_approx_eq_eps MathProgBase.getsolution(m)[2] 1.0/sqrt(2.0) 1e-6
-@test_approx_eq_eps MathProgBase.getsolution(m)[3] 1.0/sqrt(2.0) 1e-6
+@test_approx_eq_eps MathProgBase.getobjval(m) -sqrt(2.0) tol
+@test_approx_eq_eps MathProgBase.getsolution(m)[1] 1.0 tol
+@test_approx_eq_eps MathProgBase.getsolution(m)[2] 1.0/sqrt(2.0) tol
+@test_approx_eq_eps MathProgBase.getsolution(m)[3] 1.0/sqrt(2.0) tol
+if duals
+  d = MathProgBase.getconicdual(m)
+  @test_approx_eq_eps d[1] -sqrt(2.0) tol
+  @test_approx_eq_eps d[2] -sqrt(2.0) tol
+  @test_approx_eq_eps d[3] 1.0 tol
+  @test_approx_eq_eps d[4] 1.0 tol
+end
+
+end
+
+function conicEXPtest(s::MathProgBase.AbstractMathProgSolver;duals=false, tol=1e-6)
+# Problem 4 - ExpPrimal
+# min x + y + z
+#  st  y e^(x/y) <= x, y > 0 (i.e (x, y, z) are in the exponential primal cone)
+#      x == 1
+#      y == 2
+
+println("Problem 4")
+m = MathProgBase.model(s)
+MathProgBase.loadconicproblem!(m,
+[1.0, 1.0, 1.0],
+[0.0 1.0 0.0;
+ 1.0 0.0 0.0],
+[2.0, 1.0],
+[(:Zero,1:2)],
+[(:ExpPrimal, 1:3)])
+MathProgBase.optimize!(m)
+@test MathProgBase.status(m) == :Optimal
+@test_approx_eq_eps MathProgBase.getobjval(m) 6.2974 tol
+@test_approx_eq_eps MathProgBase.getsolution(m)[1] 1.0 tol
+@test_approx_eq_eps MathProgBase.getsolution(m)[2] 2.0 tol
+@test_approx_eq_eps MathProgBase.getsolution(m)[3] 3.29744 tol
+
+if duals
+  # cvx_begin
+  #     variable x, y, z;
+  #     dual variable a, b, c, d;
+  #     minimize( x + y + z );
+  #     subject to
+  #         a : x == 1;
+  #         b : y == 2
+  #         { x, y, z } == exponential( 1 );
+  # cvx_end
+  # gives
+  # a = 2.6488
+  # b = 1.8243
+  d = MathProgBase.getconicdual(m)
+  @test_approx_eq_eps d[1] 1.8243 tol
+  @test_approx_eq_eps d[2] 2.6488 tol
+end
+
+# Problem 4A - ExpPrimal
+# Same as previous, except we put :ExpPrimal on constr_cones
+println("Problem 4A")
+m = MathProgBase.model(s)
+MathProgBase.loadconicproblem!(m,
+[1.0, 1.0, 1.0],
+[0.0 1.0 0.0;
+ 1.0 0.0 0.0;
+ -1.0 0.0 0.0;
+ 0.0 -1.0 0.0;
+ 0.0 0.0 -1.0],
+[2.0, 1.0, 0.0, 0.0, 0.0],
+[(:Zero,1:2), (:ExpPrimal, 3:5)],
+[(:Free, 1:3)])
+MathProgBase.optimize!(m)
+@test MathProgBase.status(m) == :Optimal
+@test_approx_eq_eps MathProgBase.getobjval(m) 6.2974 tol
+@test_approx_eq_eps MathProgBase.getsolution(m)[1] 1.0 tol
+@test_approx_eq_eps MathProgBase.getsolution(m)[2] 2.0 tol
+@test_approx_eq_eps MathProgBase.getsolution(m)[3] 3.29744 tol
+
+if duals
+  d = MathProgBase.getconicdual(m)
+  @test_approx_eq_eps d[1] 1.8243 tol
+  @test_approx_eq_eps d[2] 2.6488 tol
+  # TODO: How do I calculate the dual of the exp constraint and check it with a non-SCS library?
+  #
+end
+end
+
+
+function conicSDPtest(s::MathProgBase.AbstractMathProgSolver;duals=false, tol=1e-6)
+
+function is_symmetric(A::Matrix)
```

Comment:
Not quite; in the example above, the answer is wrong if you replicate the
lower triangle up (because y[1,2] is no longer -10, so the optimal value is
wrong) and wrong if you replicate the upper triangle down (because then y
is not PSD).

On Thu, Jan 1, 2015 at 1:57 PM, Miles Lubin <notifications@github.com>
wrote:

> In test/conicinterface.jl
> <https://github.com/JuliaOpt/MathProgBase.jl/pull/49#discussion-diff-22402513>
> :
>
> > +@test_approx_eq_eps MathProgBase.getsolution(m)[2] 2.0 tol
> > +@test_approx_eq_eps MathProgBase.getsolution(m)[3] 3.29744 tol
> > +
> > +if duals
> > +  d = MathProgBase.getconicdual(m)
> > +  @test_approx_eq_eps d[1] 1.8243 tol
> > +  @test_approx_eq_eps d[2] 2.6488 tol
> > +  # TODO: How do I calculate the dual of the exp constraint and check it with a non-SCS library?
> > +  #
> > +end
> > +end
> > +
> > +
> > +function conicSDPtest(s::MathProgBase.AbstractMathProgSolver;duals=false, tol=1e-6)
> > +
> > +function is_symmetric(A::Matrix)
>
> Would a workaround be to replicate the lower triangle to the upper
> triangle when returning the SDP solution?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaOpt/MathProgBase.jl/pull/49/files#r22402513>.
>



-- 
Madeleine Udell
PhD Candidate in Computational and Mathematical Engineering
Stanford University
www.stanford.edu/~udell